src/upgrades/mongodb_upgrade.py

```
@@ -288,19 +307,89 @@ def is_cluster_healthy(self) -> bool:
             return False
 
         try:
-            if self.charm.is_role(Config.Role.CONFIG_SERVER) or self.charm.is_role(
-                Config.Role.SHARD
-            ):
-                # TODO Future PR - implement node healthy check for entire cluster
-                return False
             if self.charm.is_role(Config.Role.REPLICATION):
                 return self.are_replica_set_nodes_healthy(self.charm.mongodb_config)
+            else:
+                return self.are_cluster_nodes_healthy()
         except (PyMongoError, OperationFailure, ServerSelectionTimeoutError) as e:
             logger.error(
                 "Cannot proceed with upgrade. Failed to check cluster health, error: %s", e
             )
             return False
 
+    def are_cluster_nodes_healthy(self) -> bool:
+        """Returns true if all nodes in the MongoDB sharded cluster are healthy."""
+        if self.charm.is_role(Config.Role.REPLICATION):
+            return False
```

This may mask a bug along the lines if this is used by a replica set. I believe it's better to drop this check and let it crash in that case.

---

Smart - or even better merge `are_cluster_nodes_healthy` and `are_replica_set_nodes_healthy` into one function. Let me check this feasibility.